packages/explorer/src/observer/decorator.ts

```diff
@@ -15,106 +15,111 @@ export type ObserverOptions = {
+
+        const calls = args.calls
+          .map((call) => {
+            if (!call || typeof call !== "object") return undefined;
+            if (!("to" in call) || typeof call.to !== "string") return undefined;
+            if (!("functionName" in call) || typeof call.functionName !== "string") return undefined;
+            if (!("args" in call) || !Array.isArray(call.args)) return undefined;
+
+            return {
+              to: call.to as Address,
+              functionName: call.functionName,
+              args: call.args,
+              ...("value" in call && typeof call.value === "bigint" && { value: call.value }),
+            };
+          })
+          .filter(isDefined);
+        if (calls.length === 0) return write;
+
+        emit("write", {
+          writeId,
+          // TODO: type as SessionClient once available from entrykit
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          from: (client as any).userAddress ?? client.account!.address,
+          calls,
         });
-      });
-
-      return write;
-    },
-
-    async writeContract(args) {
-      const writeId = `${client.uid}-${++writeCounter}`;
-      const write = getAction(client, writeContract, "writeContract")(args);
-
-      emit("write", {
-        writeId,
-        from: client.account!.address,
-        calls: [
-          {
-            to: args.address,
-            functionName: args.functionName,
-            args: (args.args ?? []) as never,
-            ...(args.value && { value: args.value }),
-          },
-        ],
-      });
-      Promise.allSettled([write]).then(([result]) => {
-        emit("write:result", { ...result, writeId });
-      });
-
-      write.then((hash) => {
-        const receipt = getAction(client, waitForTransactionReceipt, "waitForTransactionReceipt")({ hash });
-
-        emit("waitForTransactionReceipt", { writeId, hash });
-        Promise.allSettled([receipt]).then(([result]) => {
-          emit("waitForTransactionReceipt:result", { ...result, writeId });
+        Promise.allSettled([write]).then(([result]) => {
+          emit("write:result", { ...result, writeId });
+        });
+
+        write.then((userOpHash) => {
+          const receipt = getAction(
+            client,
+            waitForUserOperationReceipt,
+            "waitForUserOperationReceipt",
+          )({ hash: userOpHash });
+
+          emit("waitForUserOperationReceipt", { writeId, userOpHash });
+          Promise.allSettled([receipt]).then(([result]) => {
+            emit("waitForUserOperationReceipt:result", { ...result, writeId });
+          });
+        });
+
+        return write;
+      },
+
+      async writeContract(args) {
+        const writeId = `${client.uid}-${++writeCounter}`;
+        const write = getAction(client, writeContract, "writeContract")(args);
+
+        emit("write", {
+          writeId,
+          from: client.account!.address,
```

Comment:
missed in #3364 but i think we should also add the `client.userAddress` check here
```suggestion
          // TODO: type as SessionClient once available from entrykit
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          from: (client as any).userAddress ?? client.account!.address,
```